engine/table/src/main/java/io/deephaven/engine/table/impl/naturaljoin/IncrementalNaturalJoinStateManagerTypedBase.java

```diff
@@ -557,8 +568,26 @@ public WritableRowRedirection buildRowRedirectionFromRedirections(QueryTable lef
                     final WritableRowSet leftRowSet = this.mainLeftRowSet.getUnsafe(ii);
                     if (leftRowSet != null && !leftRowSet.isEmpty()) {
                         final long rightRowKeyForState = mainRightRowKey.getUnsafe(ii);
-                        checkExactMatch(exactMatch, leftRowSet.firstRowKey(), rightRowKeyForState);
-                        leftRowSet.forAllRowKeys(pos -> innerIndex[(int) pos] = rightRowKeyForState);
+                        if (rightRowKeyForState == RowSet.NULL_ROW_KEY) {
```

Comment:
This is what needs replicated to `IncrementalNaturalJoinStateManagerTypedBase#buildIndexedRowRedirection`, the handling of RHS duplicates after the intial build.

There is a lot of duplication here as well.